update Sigstore community roadmap (#315)
sigstore/community: ROADMAP.md

## Brief Overview of Sigstore

[Sigstore](https://sigstore.dev) was founded in June of 2020 to address the fundamental challenges preventing wide scale usage of digital signatures within the software supply chain. Sigstore is a set of open source projects and services that dramatically simplify the creation and verification of digital signatures. It achieves this through a modular architecture that integrates proven technology solutions from adjacent domains (including Certificate Transparency, OpenID Connect, and Workload Identity among others) while reusing industry standard cryptographic algorithms and best practices. 

Sigstore is inherently flexible and can natively support a variety of:

* deployment scenarios (e.g. public-facing VS private enterprise deployments), 

> * key management approaches (e.g. hardware-backed keys stored in HSMs or Yubikeys or ephemeral scenarios where a private key can be deleted immediately after use),

Reviewer: Deleted could also be cleared from memory.

## Current Status (as of September 2023)

At the time of writing, the public Rekor signature transparency log contains over 32 million entries representing signatures from more than 17,000 unique open source projects including Kubernetes, CPython, Istio, urllib3, and Tekton (more [adopters](https://github.com/sigstore/friends) can be seen at the Sigstore [project landscape](https://landscape.openssf.org/sigstore)). Sigstore’s certificate authority has issued more than 21 million short-lived code signing certificates based on the OpenID Connect identity binding approach.  

Sigstore announced the [general availability](https://openssf.org/press-release/2022/10/25/sigstore-announces-general-availability-at-sigstorecon/) of its public-good services in October of 2022 which built upon the v1.0 releases of Cosign, Fulcio, and Rekor. The public services have consistently met their 99.5% availability SLO since that milestone.

Sigstore has several implementations of its core signing and verification workflows for popular programming languages, including:

> | Programming Language | Repository               |       Status      |

Reviewer 2: anything for sigstore-js?

|:--------------------:|--------------------------|:-----------------:|

> |        Golang        | [sigstore/cosign](https://github.com/sigstore/cosign)          |         GA        |

Reviewer 3: Imo it's best to refer to the Go programming language as "Go" and leave "Golang" for use as a search term. 😄

> |        Golang        | [sigstore/cosign](https://github.com/sigstore/cosign)          |         GA        |

Reviewer: ```suggestion
|        Go        | [sigstore/cosign](https://github.com/sigstore/cosign)          |         GA        |
```

|        Python        | [sigstore/sigstore-python](https://github.com/sigstore/sigstore-python) |         GA        |
|      Javascript      | [sigstore/sigstore-js](https://github.com/sigstore/sigstore-js)     |         GA        |
|         Java         | [sigstore/sigstore-java](https://github.com/sigstore/sigstore-java)   |        Beta       |
|         Rust         | [sigstore/sigstore-rs](https://github.com/sigstore/sigstore-rs)     |        Beta       |

> |        Golang        | [sigstore/sigstore-go](https://github.com/sigstore/sigstore-go)     | Under development |

Reviewer: ```suggestion
|        Go            | [sigstore/sigstore-go](https://github.com/sigstore/sigstore-go)     | Under development |
```

* Provide tooling for individuals to monitor logs for their signing identities

### **Collaborate with academic community to provide stronger trust assurances and privacy guarantees**

* Provide attestations of trusted actions, such as verifying identity tokens, to minimize needed trust in service operators
* Use modern cryptography (zero-knowledge proofs, verifiable random functions) to remove PII where possible

### **Ensure that verification is as seamless as signing**

* Integrate with existing verification policy languages (e.g. Rego, CUE)
* Provide a reference implementation for how to verify signatures across various points in the software development lifecycle

### **Offer intuitive and ergonomic tooling for engineers and cloud operations**

* [Cosign CLI](https://github.com/sigstore/cosign) is the official Sigstore CLI multi-tool for signing and verification in a variety of environments and use cases.
* [sigstore-policy-controller](https://github.com/sigstore/policy-controller) is the center of Sigstore-oriented verification flows in Kubernetes environments.

# Roadmap

## Client SDKs

### Short Term (< 6 mo)

> * Releases:

Reviewer 2: anything for sigstore-js here?

Reviewer: @bdehamer Anything you wanted to call out? Any major releases in the next year?

Reviewer 4: No major releases planned. I've got a long list of small features/improvements I'll be working on but nothing worth calling-out here.

### Medium Term (6-12 mo)

* FFI Implementation
* Port at least one language to make use of FFI implementation
* Releases:
    * sigstore-go v1 release
    * Gitsign v1 release
    * policy-controller v1 release
* Supporting BYO PKI and private deployments in Sigstore clients
* Improved testing
    * End-to-end testing
    * Conformance tests are passing for all clients
    * Fuzz tests
* Improving verification policies
    * Rego/CUE integration
    * [Hammurabi from web PKI](https://www.andrew.cmu.edu/user/bparno/papers/hammurabi.pdf)
* Reduce size of Cosign binary, fewer dependencies
* Privacy-preserving telemetry (https://divviup.org/)

### Long Term (12+ mo)

* Plumbing and porcelain model for Cosign, to improve Cosign's user experience
* Policy: sophisticated verification policies, or deeper integration with a policy evaluation service
* Further integration with TUF through delegations or trusted verification material distribution, and in-toto for signed provenance
* Adding additional conformance tests to verify consistent behavior across clients, including negative tests
* Bundle format includes attestation support that interops with SLSA

> * Reference implementation of how to use signature verification across various points in the software development lifecycle, leveraging policy-controller as one example

Reviewer: ```suggestion
* Reference implementation of how to use signature verification across various points in the software development lifecycle, leveraging policy-controller as one example
* Tooling to support client revocation policies
```

## Signature Transparency Log (Rekor)

### Short Term (< 6 mo)

> * Add 2 independent log monitors

Reviewer 3: Should we call out what we want to use to monitor the logs? I know we have a monitoring project here in the Sigstore org that I believe powers the Purdue monitor. Should there be a roadmap for that kind of functionality?

Reviewer: Good suggestion, I’ll add a section on that. I have a roadmap for what’s needed to enable witnessing and identity monitoring.

Reviewer: Added roadmap, PTAL!

    * Monitors will only verify log consistency
* Witnessing
    * Monitors should transition into fully fledged witnesses
    * Witnesses should publish signed checkpoints to distributor
* V1 for GitHub Actions-based identity monitor
* Timestamp at Entry Upload
    * Timestamp should be provided by TSA
    * Timestamp becomes immutable and verifiable
* Complete architecture document

> * Publish inclusion events to pub/sub queueo

Reviewer: ```suggestion
* Publish inclusion events to pub/sub queue
```

    * Must support witnessing without compromises
* MVP split-view attack mitigation
    * Client requests witnessed checkpoints from a set of witnesses or distributors
    * Client verifies inclusion proof from log using checkpoint size
    * Offline verifiable proofs include witnessed checkpoints in Bundle
* Verifiable Search Design
    * Detecting credential compromise without requiring constant log scraping
* Design transparent redaction mechanism
* Reviewing sharding strategy to better align with Certificate Transparency's sharding
* Experiment with serverless/log tile format for caching and scalability
* Design log replication mechanism for scaling instances (possibly can use serverless)

### Long Term (12+ mo)

* Verifiable Search Implementation
* Package repositories participate in mitigating split-views
    * Witnessed checkpoints attached to artifact signature metadata
    * Repositories verify log consistency
* Implementation for log replication
* Transparent redaction implementation
* Privacy-preserving log queries, where the log is not aware of client's artifact of interest
* Roughtime for distributed trusted timestamps
* Continued fuzzing of large API surface

> 

Reviewer: ```suggestion

## Transparency Log Monitor

### Short Term (< 6 mo)

* Initial launch of Sigstore's [log monitor on GitHub Actions](https://github.com/sigstore/rekor-monitor)
   * Monitors acts as both a witness (co-signing checkpoints) and Verifier (monitoring given identities)
   * Co-signed checkpoints are published to GitHub
   * Verifier monitors for OIDC identities and key fingerprints
   * Alerts are filed as GitHub issues
   * Supports sharded log
* At least 2 independent instances of GitHub Actions-based log monitor running and publicly documented
* Publish blog posts on monitoring requirements as an artifact producer
* Create log cloning tool to support offline log monitoring and analysis
   * Can also be used to transform an existing log into a serverless layout

### Medium Term (6-12 mo)

* V1 launch of Sigstore's [log monitor on GitHub Actions](https://github.com/sigstore/rekor-monitor)
   * Witnessed checkpoints are published to distributors, operated by Sigstore and/or community members
   * Support monitoring Fulcio's CT log
   * Alerts are sent in a variety of ways (e.g. email)
   * Productionized and resilient against intermittent failures
* Publish select monitors to status.sigstore.dev

### Long Term (12+ mo)

* Optionality around running witness in different environments, e.g. hardware or cloud
* Integration with other efforts in transparency ecosystems
   * Publishing checkpoints to external community-operated distributors
* Collaborate with CT log monitors (e.g. [sslmate](https://sslmate.com/certspotter/)) to improve identity monitoring

```

Reviewer: @trevrosen PTAL and let me know if you see anything that's missing.

Reviewer 3: Looks great - thanks!

## Certificate Authority (Fulcio)

### Short Term (< 6 mo)

* Hashed email addresses/PII
    * Still supports monitoring identities
* Complete architecture document
* Fulcio CI identity provider onboarding automation
* Standardize OIDC identity provider requirements for inclusion in public instance

### Medium Term (6-12 mo)

* Privacy-preserving identities
    * VRFs or ZKP to prevent attacker from learning when a signing event occurred for a given identity
    * Must still support identity monitoring
* Zero-knowledge proofs that prove a CA verified an identity token

### Long Term (12+ mo)

* ACME for code-signing certificates (ongoing RFC)
* Native SPIFFE support with X509-SVIDs as authenticators, rather than supporting only OIDC

> * Support revocation for private deployments

Reviewer 5: is revocation for the public deployment intentionally omitted? This could be useful for key or account compromise (discussed in #sigstore/cosign#677)

Reviewer: It's a hard problem on a large scale - How do we authenticate requests to revoke content? How do we maintain an indefinitely growing list? WebPKI doesn't deal with revocation well, and I worry we'd run into the same issues.

Reviewer 5: I agree that this is challenging, and should be part of the 12 months+ roadmap. But if Rekor is used to discover a mis-behaving key/identity, Sigstore should do *something* about this. Whether this is a revocation list, TUF delegations to trusted identities, or some other mechanism.

Reviewer 6: Is this revoking specific entries in the log by UUID, revoking everything signed with public key X, revoking everything signed by x509 cert with serial Y, or revoking everything signed by `foo@bar.com` between a date range?

Reviewer 6: I that case I agree. I think it is up to the consumers to determine what is trustworthy. 

I imagine package manager services would manage this in most cases so it would be transparent to the user, and the operators of each package service can handle authz for determining who can revoke trust.

Reviewer 5: I agree that the policy should be up to consumers, but Sigstore can provide tooling to make these revocations possible

Reviewer 7: @mnm678 Could the PKI for public deployments (including policies) be covered by [Line 380](https://github.com/sigstore/community/pull/315/files#diff-683343bdf93f55ed3cada86151abb8051282e1936e58d4e0a04beca95dff6e51R380)?

Reviewer 5: possibly, depending on how that is implemented

Reviewer: Do you have suggestions of wording changes to either this bullet point or the one Trishank noted that would cover what you're looking for? A mention of "client support for revocation policies"?

Reviewer 5: maybe "tooling to support client revocation policies"

Reviewer 7: I'd say it's more general than just revocation, but something along the lines of: "A solution for distributing, revoking, and rotating public keys and policies used to verify public artifacts"

Reviewer: I think that fits under the client roadmap with "Policy: sophisticated verification policies, or deeper integration with a policy evaluation service".

Reviewer 8: I suggest we don't use the word "revocation" until we had a discussion about it. Avoiding any "revocation" solution is generally a good thing, since most of them has not worked very well. So Hayden's suggestion seems good.

Reviewer 7: Right, that makes sense: each, say, package registry may wish to handle revocation separately from Sigstore. Sigstore should ideally "just" be a (but not the only) source of truth.

Reviewer 7: Do we have a resolution about this? @mnm678 @haydentherapper

Reviewer 5: Hayden's interpretation of it fitting under the policy item works for me

## Deployment Infrastructure / Configuration

### Short Term (< 6 mo)

* Move builds off GCB onto GHA
* Adopt SLSA L3 builders and publish provenance

### Medium Term (6-12 mo)

* Refactoring helm-charts based on new common chart format
* Work with the community to support private deployments

### Long Term (12+ mo)

* Terraform for non-GCP architecture

## Documentation

### Short Term (< 6 mo)

> * Documentation restructured around user journeys

Reviewer: ```suggestion
* Documentation restructured around user journeys
* Document workload identity/automated signing in addition to developer-driven signing
* Terminology page
```

### Medium Term (6-12 mo)

* Developer documentation
* Architecture documents published

### Long Term (12+ mo)

* Documentation versioned for tooling releases
* Improvements to branding website

## Public-Good Instances

### Short Term (< 6 mo)

* Publish document on public vs private deployments
* Maintain operational excellence:
    * 99.5% SLO
    * Multi-vendor on call rotation
    * Continued burn down of open tickets
    * Periodic DiRT testing
* Perform 1 log sharding and log key rotation

> * Complete architecture document, get commitment on operational transfer

Reviewer: ```suggestion
* Complete architecture document, get commitment on operational transfer
* Migrate TUF root repository over to TUF-on-CI
```

### Medium Term (6-12 mo)

* Begin transition of public instance to neutral third party
* Release automation
    * New releases should go to staging automatically
    * Releases should seamlessly move from staging to production with automated testing
* New bleeding-edge environment

>     * Hourly/nightly releases should automatically deploy to this env

Reviewer: That's a good callout, we have an issue in the private infra repo for tracking automatically filing issues on GH, but this has not been completed.

* Availability improvements
    * Regionalization of components

>     * Read replicas of log

Reviewer: ```suggestion
    * Read replicas of log
* Improve automated issue filing for alerts
```

### Long Term (12+ mo)

> * Public root of trust trusted by private ecosystems (Apple App Store, Google Play Store, Windows)

Reviewer: Policy would need to be a part of this, yes.

* Public instance has been fully transitioned to neutral third party

## Community

### Short Term (< 6 mo)

* Add more community-operated timestamp authorities
* TUF root supports additional delegations

### Medium Term (6-12 mo)

* BigQuery dataset for researchers
* Alignment within OpenSSF (SCI WG, Security Toolbelt)

### Long Term (12+ mo)

> * Sigstore API specified in IETF RFC

Reviewer: ```suggestion
* Sigstore API specified in IETF RFC

## Timestamp Authority

Feature complete. Future improvements include integrating timestamps into Rekor entries and exploring Roughtime as an alternative for distributed trusted timestamping.
```

1. Develop language-specific implementation of Sigstore’s signing and verification workflows (if one does not already exist)
2. Integrate language-specific Sigstore signing and verification functionality into package management tooling
3. Add support for storing Sigstore signatures & attestations in the package registry for consumers to access
4. Publish community RFC proposal describing end-to-end workflows, ecosystem-specific threat model, and roadmap for adoption
5. Once RFC proposal is approved per community norms, drive implementation plan
6. Work to drive initial package adoption strategy; this often involves highly visible or valuable packages that can serve as references for other package maintainers within the ecosystem
7. Work with popular build & packaging extensions (e.g. GoReleaser, JReleaser) and builder templates (Jenkins plugins, GitHub Actions, 
8. Release all code & service extensions required for Sigstore support in a fully-supported mode
9. Presuming success to this point, set date for mandating Sigstore signing and verification for all packages in ecosystem 

## Current Ecosystem Adoption Status

> The npm registry, the world's largest open source package manager for the Node.js ecosystem, [recently announced built-in support for Sigstore-signed SLSA build provenance](https://github.blog/2023-04-19-introducing-npm-package-provenance/) to users. This allows consumers of packages from npm to independently verify the link between a package fetched from the npm registry to the source code from which it was built. The build provenance attestation is cryptographically signed using Sigstore.

Reviewer 9: ```suggestion
The npm registry, the world's largest open source package manager for the JavaScript ecosystem, [recently announced built-in support for Sigstore-signed SLSA build provenance](https://github.blog/2023-04-19-introducing-npm-package-provenance/) to users. This allows consumers of packages from npm to independently verify the link between a package fetched from the npm registry to the source code from which it was built. The build provenance attestation is cryptographically signed using Sigstore. Other package managers may find the [documentation of npm's Sigstore adoption and resulting design decisions](https://repos.openssf.org/build-provenance-for-all-package-registries) useful.
```
